File: packages/stera_recorder/ios/stera_recorder/Sources/stera_recorder/encoding/FrameSamplerImpl.swift
import Foundation

/// Deterministic timestamp-based RGB frame sampler.
class FrameSamplerImpl: FrameSampler {

    private var targetFps: Int
    private var encodeFrameIntervalNs: Int64
    private var sessionStartTimestampNs: Int64 = Int64.min
    private var deterministicFrameIndex: Int64 = 0
    private var totalExpectedFrameCount: Int64 = 0
    private var cumulativeTrackingPauseNs: Int64 = 0

    init(targetFps: Int = 30) {
        self.targetFps = max(targetFps, 1)
        self.encodeFrameIntervalNs = Int64(1_000_000_000 / self.targetFps)
    }

    func configure(targetFps: Int) {
        self.targetFps = max(targetFps, 1)
        self.encodeFrameIntervalNs = Int64(1_000_000_000 / self.targetFps)
    }

    func shouldEncodeFrame(timestampNs: Int64) -> Bool {
        if sessionStartTimestampNs == Int64.min {
            sessionStartTimestampNs = timestampNs
            deterministicFrameIndex = 0
            totalExpectedFrameCount = 0
        }
        let elapsedNs = max(timestampNs - sessionStartTimestampNs - cumulativeTrackingPauseNs, 0)
        let expectedFramesByNow = (elapsedNs / encodeFrameIntervalNs) + 1

        if expectedFramesByNow > totalExpectedFrameCount {
            totalExpectedFrameCount = expectedFramesByNow
        }
        if expectedFramesByNow <= deterministicFrameIndex {
            return false
        }
        deterministicFrameIndex = expectedFramesByNow
        return true
    }

    func recordTrackingPauseOffset(offsetNs: Int64) {
        cumulativeTrackingPauseNs += max(offsetNs, 0)
    }

    func reset() {
        sessionStartTimestampNs = Int64.min
        deterministicFrameIndex = 0
        totalExpectedFrameCount = 0
        cumulativeTrackingPauseNs = 0
    }

    func getDriftStats() -> [String: Any] {
        return [
            "sessionStartTimestampNs": sessionStartTimestampNs,
            "deterministicFrameIndex": deterministicFrameIndex,
            "totalExpectedFrameCount": totalExpectedFrameCount,
            "cumulativeTrackingPauseNs": cumulativeTrackingPauseNs,
            "targetFps": targetFps,
            "encodeFrameIntervalNs": encodeFrameIntervalNs
        ]
    }
}

File: packages/stera_recorder/android/src/main/kotlin/open/fpvlabs/stera/ar_recorder/encoding/FrameSamplerImpl.kt
package open.fpvlabs.stera.ar_recorder.encoding

/**
 * Deterministic timestamp-based RGB frame sampler.
 */
class FrameSamplerImpl(
    private var targetFps: Int = 30
) : FrameSampler {
    private var encodeFrameIntervalNs: Long = 1_000_000_000L / targetFps
    private var sessionStartTimestampNs: Long = Long.MIN_VALUE
    private var deterministicFrameIndex: Long = 0L
    private var totalExpectedFrameCount: Long = 0L
    private var cumulativeTrackingPauseNs: Long = 0L

    override fun configure(targetFps: Int) {
        this.targetFps = targetFps.coerceAtLeast(1)
        this.encodeFrameIntervalNs = 1_000_000_000L / this.targetFps
    }

    override fun shouldEncodeFrame(timestampNs: Long): Boolean {
        if (sessionStartTimestampNs == Long.MIN_VALUE) {
            sessionStartTimestampNs = timestampNs
            deterministicFrameIndex = 0L
            totalExpectedFrameCount = 0L
        }
        val elapsedNs = (timestampNs - sessionStartTimestampNs - cumulativeTrackingPauseNs).coerceAtLeast(0L)
        val expectedFramesByNow = (elapsedNs / encodeFrameIntervalNs) + 1L
        if (expectedFramesByNow > totalExpectedFrameCount) {
            totalExpectedFrameCount = expectedFramesByNow
        }

        if (expectedFramesByNow <= deterministicFrameIndex) {
            return false
        }
        deterministicFrameIndex = expectedFramesByNow
        return true
    }

    override fun recordTrackingPauseOffset(offsetNs: Long) {
        cumulativeTrackingPauseNs += offsetNs.coerceAtLeast(0L)
    }

    override fun reset() {
        sessionStartTimestampNs = Long.MIN_VALUE
        deterministicFrameIndex = 0L
        totalExpectedFrameCount = 0L
        cumulativeTrackingPauseNs = 0L
    }

    override fun getDriftStats(): Map<String, Any> {
        return mapOf(
            "sessionStartTimestampNs" to sessionStartTimestampNs,
            "deterministicFrameIndex" to deterministicFrameIndex,
            "totalExpectedFrameCount" to totalExpectedFrameCount,
            "cumulativeTrackingPauseNs" to cumulativeTrackingPauseNs,
            "targetFps" to targetFps,
            "encodeFrameIntervalNs" to encodeFrameIntervalNs
        )
    }
}
